src/pushsource/_impl/backend/koji_source.py

```diff
@@ -509,6 +521,94 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all image archives which are not containers (tar)
+        vmi_archives = [
+            elem
+            for elem in archives
+            if elem.get("btype") == "image" and elem.get("type_name") != "tar"
```

Comment:
That would much even qcow2 `'type_extensions': 'qcow2'`
Do we want to support it or just: "raw", "raw-xz", "vhd", "vhd-compressed"
IMHO it is better to support just known extensions and expends it later if needed.
@rohanpm  WDYT?

Reply:
We can do it or leave it open.. By doing it we'll need to include GCP later on and adjust the library for each new "image type" that appears, which can be a bit cumbersome IMO.

By not filtering we allow all image types and may filter them even more inside pubtools, but up to you guys, whatever works best 😃